### PR TITLE
feat(orchestrator): fix pagination issue and add filters to instances and overviews

### PR DESCRIPTION
**Pull request by gciavarrini**
Modify the definitions of `/v2/workflows/overview` and `/v2/workflows/instances` to properly handle pagination and to use filters.

Add unit tests for `queryBuilder` and `DataIndexService` to test the pagination and filter functionalities.

### CODE REVIEW

**Comment by batzionb** [comment]
@gciavarrini 
PTAL one of the unit tests is failing

https://github.com/janus-idp/backstage-plugins/actions/runs/10253689160/job/28366890785?pr=2007#step:3:260

File: plugins/orchestrator-backend/src/service/api/v2.ts [hunk 1]
```diff
@@ -75,19 +78,21 @@ export class V2 {
...
     const totalCount =
       await this.orchestratorService.fetchInstancesTotalCount();
 
     const result: ProcessInstanceListResultDTO = {
       items: instances?.map(mapToProcessInstanceDTO),
       paginationInfo: {
-        pageSize: pagination.limit,
-        page: pagination.offset,
+        pageSize: pagination?.limit,
+        page: pagination?.offset,
```
Comment (batzionb) [inline plugins/orchestrator-backend/src/service/api/v2.ts:95]: this should be pagination?.offset / pageSize

File: plugins/orchestrator-backend/src/types/pagination.ts [hunk 2]
```diff
@@ -8,23 +10,33 @@ export interface Pagination {
...
+  if (!req.body?.paginationInfo) {
+    return pagination;
   }
+  const { page, pageSize, orderBy, orderDirection } = req.body
+    .paginationInfo as PaginationInfoDTO;
 
-  if (!isNaN(Number(req.query.pageSize))) {
-    pagination.limit = Number(req.query.pageSize);
+  if (!isNaN(Number(page))) {
+    pagination.offset = Number(page);
```
Comment (batzionb) [inline plugins/orchestrator-backend/src/types/pagination.ts:27]: To fix an already existing bug,
This should be
pagination.offset = Number(page) * Number(pageSize)

File: plugins/orchestrator-backend/src/types/filters.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      operator,
+      fieldValue: parseFilterValue(fieldValue),
+    };
+  }
+
+  return undefined;
+}
+
+function parseFilterValue(value: FilterValue): string | number | boolean {
+  if (typeof value === 'boolean') return Boolean(value);
```
[the remark this answers is not in the document]
Comment (gciavarrini) [reply plugins/orchestrator-backend/src/types/filters.ts:36]: > if the type is boolean Boolean(value) equals to value`
correct. I'll fix it
>  if type is string ${value} equals to value
no, this method build the filter for graphQL query. I need to render the string properly (that is `"mystring"` with quotes)
>  as to the third condition - how can a value be also boolean and also number?
it's a typo. I'll replace `boolean` with `number`. thanks for catching this! :D

[the remark this answers is not in the document]
Comment (batzionb) [reply plugins/orchestrator-backend/src/types/filters.ts:36]: no, this method build the filter for graphQL query. I need to render the string properly (that is "mystring" with quotes)
right
sorry, I missed the quotes

File: plugins/orchestrator-backend/src/types/pagination.ts (hunk 2 above)
Comment (gciavarrini) [reply plugins/orchestrator-backend/src/types/pagination.ts:27]: what if `pageSize` is not defined?

File: plugins/orchestrator-backend/src/service/api/v2.ts (hunk 1 above)
Comment (gciavarrini) [reply plugins/orchestrator-backend/src/service/api/v2.ts:95]: ok, now i get it. Thanks for clarifying.
I'll change from `page` to `offset` to be, also,  "compliant" with graphql syntax.

File: plugins/orchestrator-backend/src/types/filters.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (gciavarrini) [reply plugins/orchestrator-backend/src/types/filters.ts]: addressed

File: plugins/orchestrator-backend/src/service/api/v2.ts (hunk 1 above)
Comment (gciavarrini) [reply plugins/orchestrator-backend/src/service/api/v2.ts]: done

**Review by batzionb — APPROVED** [review]

**Pull request merged**